core/git_command.py

```diff
@@ -150,19 +156,66 @@ def raise_error(msg):
+                    util.log.panel("> {}\n{}\n{}".format(command_str, stdout, stderr))
+                else:
+                    util.log.panel("{}\n{}".format(stdout, stderr))
 
+        if not p.returncode == 0 and throw_on_stderr:
+            # this piece of code detect git error meesages related to authentification.
+            # stdin must be None, because we will use stdin to pass username and password from Sublime to Git.
+            if catch_credential_errors and stdin is None and 'fatal: could not read Username for' in stderr:
+                command = [self.git_binary_path, '-C', working_dir or self.repo_path] + list(arg for arg in args if arg)
+                command_str = " ".join(command)
+
+                self._stdout = None
+                is_window_command = hasattr(self, "window")
+                window = self.window if is_window_command else self.view.window()
+                window.show_input_panel('username:', '', lambda username: self.on_username(window, command_str, username, on_result), None, self.on_cancel)
+
+                # wait that the Git command has been executed in the show_input_panel thread.
+                if is_window_command:
+                    while self._stdout is None:
+                        time.sleep(1)
+                    stdout = self.decode_stdout(stdout, savvy_settings) if decode else self._stdout
+                    return stdout
+
+                # gwenzek: HACK: for TextCommand, it seems that the wait blocks the execution of `show_input_panel`.
+                # The best solution I found is to return directly, but the output message that may be important isn't returned.
+                # But it should work in most cases, since most commands that require the output doesn't require autenthification (log, status, ...)
+                # Anyway, the output will still be shown to users if they ask for it in "show_panel_for" setting.
+                else:
+                    print('returning without waiting for stdout')
+                    return None
+
+        on_result(p.returncode, stdout, stderr)
         return stdout
 
+    def on_username(self, window, command_str, username, on_result):
+        """ Asks for the password, before calling the given command. """
+        window.show_input_panel(username + "'s password:", '', lambda password: self.on_credentials(command_str, username, password, on_result), None, self.on_cancel)
+
+    def on_credentials(self, command_str, username, password, on_result):
+        """ Launch the given command, stores the ouput and return code in self. """
+        # pexpect runs the given command in a subprocess, parses the output and feeds the username and password when asked by Git.
+        # [pexpect](https://github.com/pexpect/pexpect)
+        stdout, return_code = pexpect.run(command_str, events={'Username for': username + '\n', 'Password for': password + '\n'}, withexitstatus=True)
```

Comment:
This is the only part of the code where I use "pexpect", if someone manage to use Popen for doing the same thing, we could drop the dependency.